### ISSUE TITLE
False positive when updating React state with functional setState and object spread (prevState)

### ISSUE THREAD

**Issue by dev-gabriel-henrique** [microsoft/TypeScript#63042]
### Acknowledgement

- [x] I acknowledge that issues using this template may be closed without further explanation at the maintainer's discretion.

## Description

When using React state updates with the functional form of `setState`, TypeScript allows extra properties to be added to the returned object without reporting any type error. This results in a **false negative**, where invalid state keys silently pass type checking.

This commonly happens when spreading `prevState` and adding a property that does not exist in the state type.

## Example

```ts
type State = {
  states: string[];
  page: number;
};

const [filters, setFilters] = useState<State>({
  states: [],
  page: 1,
});

setFilters(prev => ({
  ...prev,
  state: "SP", // ❌ This property does NOT exist in `State`
}));
```

## Expected behavior

A type error should be reported when returning an object with properties that do not exist on the state type, even when using the functional setState(prev => ...) pattern.

## Environment

React with useState (this happens in versions: 18 and 19);

TypeScript (strict mode enabled);

ESLint v9 (Flat Config);

@typescript-eslint;

eslint-plugin-react-hooks;

TypeScript does not report any error in this scenario, even though state is not part of the State type.

## ESLint configuration

```ts
export default defineConfig([
  {
    files: ["**/*.{ts,tsx,js,jsx}"],
    plugins: {
      react: reactPlugin,
      "@typescript-eslint": typescriptPlugin,
      "react-hooks": reactHooksPlugin,
    },
    languageOptions: {
      parser: "@typescript-eslint/parser",
      parserOptions: {
        project: "./tsconfig.eslint.json",
      },
    },
    rules: {
      "react-hooks/rules-of-hooks": "error",
      "@typescript-eslint/no-unused-vars": "warn",
    },
  },
]);
```

Anybody can help me with this?

**Comment by MartinJohns** [comment]
Duplicate of #47758 / #241.

**Comment by dev-gabriel-henrique** [comment]
> Duplicate of [#47758](https://github.com/microsoft/TypeScript/issues/47758) / [#241](https://github.com/microsoft/TypeScript/issues/241).

This error seems to be related to React’s setState when it relies on the previous state (prevState).
Since it only occurs when React and TypeScript are used together, the question is whether this should be considered a React issue or a TypeScript one.

**Comment by MartinJohns** [comment]
> This error seems to be related to React’s setState when it relies on the previous state (prevState).

It's a duplicate of #47758, as I mentioned.

`setFilters()` requires an argument that is typed `(prevState: State) => State`. You pass in an anonymous function, and because of #47758 / #241 TypeScript does **not** use `State` as the return type of your anonymous function, but instead infers the type. The type is then checked for assignability, and additional properties are allowed in TypeScript.

You can work around this by adding an explicit type annotation to your function: `setFilters((prev): State => ({ ...prev, state: "SP" }))`

**Comment by dev-gabriel-henrique** [comment]
That makes sense, and thanks for the clarification. This is working, its not an error but it bothers me,
is there any other way to avoid this false positive without explicitly declaring (prevState): State => State?

If not, its ok for me

Thanks for answering me.
